l0rda

l0rda opened the pull request:
Я так понял задание, что Х процентов символов в массиве DNA заменяются
на новые(случайные), а не перемешиваются.

l0rda committed:
initital import

Я так понял задание, что Х процентов символов в массиве DNA заменяются
на новые(случайные), а не перемешиваются.

l0rda committed:
исправление сравнения указателей вместо объектов

l0rda committed:
исправление - гарантированная мутация

удаляем текущий нуклеотид из базового массива для замены, чтобы
нуклеотид в позиции был гарантированно заменен на другой

l0rda committed:
исправление к предыдущему коммиту

исправление к предыдущему коммиту, кое-что таки забыл :)

l0rda committed:
убрал лишнюю проверку

убрал лишнюю проверку, как подсказали в комментариях

l0rda committed:
учлучшение - замена нуклеотида в последовательности

замена нуклеотида в последовательности, но без создания нового массива

hmmvot commented on DNA/Cell.m:
```diff
@@ -0,0 +1,40 @@
+    self = [super init];
+    if (self){
+        _dnkArrayLength = 100;
+        _nucleotides = [NSMutableArray arrayWithObjects: @"A", @"T", @"G", @"C", nil];
```
Вы создаете переменные _dnkArrayLength и _nukleotides в каждом объекте класса  Cell, но храните в них одни и те же значения. Почему бы не сделать их константами или статическими переменными?

l0rda replied:
сделал их readonly

aspcartman replied:
Это не улучшило положение :)

l0rda replied:
Если можно, объясните пожалуйста, как их сделать константами или статическими переменными, но честно говоря я не вижу ничего зазорного в том, что эти данные хранятся в свойствах.

aspcartman replied:
Зазорность в том, что эти поля не уникальны для каждого из обьектов, а общие для всех. Значит они должны быть не полем обьекта, а лежать где-то в классе. А сделать очень легко - можно, например, прямо под `@implementation` положить 

``` obj-c
static NSString *nucleotides[4] = {@"A",@"T",@"G",@"C"};
```

:)

hmmvot replied:
Только учтите, что static переменная видна только в пределах того блока, в котором она была объявленна (максимум - в пределах одного файла).
Вообще советую познакомиться со спецификаторами static и extern и с директивой #define языка C, если вы не знакомы. И с самим языком тоже.

В пример от @aspcartman было бы круто добавить еще ключевое слово const после static.)

l0rda replied:
static NSString *nucleotides[4] = {@"A",@"T",@"G",@"C"};

ну вот это будет плохо, так как не доступно в категории mutator, только если через дефайн делать.

hmmvot replied:
С _dnkArrayLength так и надо поступить, а вот дефайнить массив - гиблое дело.

Вообще есть обходные пути: можете создать метод, который будет возвращать значение статической переменной, или использовать спецификатор extern.

aspcartman replied:
Я думаю, в рамках нашего курса лучшим решием было бы

``` obj-c
static const NSArray *nucleotides = @[ @"A", ... ];
```

плюс extern или соответствующий метод `- (NSArray*) possibleNucleotides`. Тогда и дефайна размера не надо и можно использовать метод `count`, который на самом то деле достаточно дешевый.

l0rda replied:
исправил все в последнем коммите, нуклеотидеы сделал через метод - (NSArray*) possibleNucleotides
для переменной DNA переписал сеттер, чтобы невозможно было переопределить переменную извне, возможно сделал это немного криво :)
с extern пробовал разобраться, но не смог пока.

hmmvot commented on DNA/Cell.h:
```diff
@@ -0,0 +1,20 @@
+}
+
+@property int dnkArrayLength;
+@property NSMutableArray *nucleotides;
```
Использование свойств, в данном случае, не оправданно.

l0rda replied:
почему нет? сделал их readonly в последнем коммите, они используются в категории

aspcartman replied:
Почему бы и нет? :)

hmmvot replied:
1) Странно создавать геттер для переменной, которая используется только внутри класса. Все таки (на мой взгляд) property нужно использовать для организации доступа извне, а это обычные переменные.
2) Опять же, на мой взгляд, это константы, а не переменные. Да еще и общие для всех объектов класса.
3) readonly в @property NSMutableArray *nucleotides не запретит изменение массива (запретит только изменение указателя на массив).

hmmvot commented on DNA/Cell.m:
```diff
@@ -0,0 +1,40 @@
+    }
+    return distance;
+}
+-(void)printDNA {
```
Вместо этого можно переопределить метод

``` objective-c
- (NSString *) description;
```

При вызове

``` objective-c
NSLog(@"%@", obj);
```

в лог печатается результат посылки сообщения description объекту obj.

l0rda replied:
сделал

hmmvot commented on DNA/main.m:
```diff
@@ -0,0 +1,69 @@
+
+@implementation Cell(mutator)
+-(void)mutate: (int) mutatePercent {
+    // mutatePercent - не может быть больше 100%
```
Скажите об этом так, чтобы программа дальше не выполнялась, а не тихонько ничего не делала :)

``` objective-c
assert(mutatePercent > 100);
```

l0rda replied:
сделал

aspcartman replied:
`NSAssert(mutatePercent > 100 || mutatePercent < 0)` :)

hmmvot commented on DNA/Cell.m:
```diff
@@ -0,0 +1,40 @@
+        for (int x=0;x<100;x++) {
+            int nucleotides_index = arc4random() % [_nucleotides count];
+            id newNucleotide = _nucleotides[nucleotides_index];
+            [DNA insertObject:newNucleotide atIndex:x];
```
``` objective-c
[DNA addObject: _nucleotides[arc4random() % [_nucleotides count]]];
```

Так быстрее, короче и проще.

l0rda replied:
это уже исправлено было

hmmvot commented on DNA/main.m:
```diff
@@ -0,0 +1,69 @@
+    if (mutatePercent > 100) {
+        return;
+    }
+    
```
Зачем выполнять написанный ниже код, если mutatePercent == 0 ?

l0rda replied:
поправил

hmmvot commented on DNA/main.m:
```diff
@@ -0,0 +1,69 @@
+        id newNucleotide = currentNuke;
+        while ([newNucleotide isEqual: currentNuke]) {
+            newNucleotide = self.nucleotides[arc4random() % [self.nucleotides count]];
+        }
```
Неприятная конструкция. Можно обойтись без подобного цикла.

l0rda replied:
как, например?
имхо это лучше, чем делать копию массива self.nucleotides, что я делал в более раннем коде.

hmmvot replied:
Попытайтесь придумать решение, которое сможет обойтись тыкания нуклеотидов наугад. Вообще это, конечно, не критично в данном случае, но поискать решение можно и просто для тренировки.

hmmvot commented on DNA/main.m:
```diff
@@ -0,0 +1,69 @@
+#import <Foundation/Foundation.h>
+#import "Cell.h"
+
+@interface Cell(mutator)
```
Категорию лучше вынести в отдельный файл (в Xcode есть специальный шаблон для этого).

l0rda replied:
вынес, не знал об этом, сам удивлялся зачем категории в основном коде

hmmvot commented:
Сразу полез в changed files, так что просмотрел некоторые уже существующие комментарии и многое надублировал. Извиняюсь.)

l0rda committed:
улучшение

категория вынесена в отдельный файл
assert на процент больше 100 и меньше 1
метод printDNA переделан, теперь переопределяется description
свойства длин днк и массив с нуклеотидами теперь readonly

l0rda committed:
улучшение

более идеологически верный NSAssert

l0rda committed:
улучшение

указывается размер массива при инициализации

l0rda committed:
улучшение

длины массива DNA и массива с нуклеотидами вынесены в define,
в объект Cell добавлен метод, который всегда возвращает массив с
нуклеотидами,
сеттер setDNA переопределен и теперь невозможно перезаписать DNA извне

hmmvot commented on DNA/Cell.h:
```diff
@@ -0,0 +1,22 @@
+}
+-(NSArray *) possibleNucleotides;
+-(void) setDNA: var;
+-(NSString *) description;
```
Не нужно объявлять этот метод здесь - он уже объявлен в NSObject.

l0rda replied:
согласен, исправил

hmmvot commented on DNA/Cell.h:
```diff
@@ -0,0 +1,22 @@
+    NSMutableArray *DNA;
+}
+-(NSArray *) possibleNucleotides;
+-(void) setDNA: var;
```
Не понял зачем это. Ваша переменная NSMutableArray *DNA и так "спрятана", для нее не существует сеттера. Зачем его определять?

l0rda replied:
Вы же сами писали:
В данный момент переменная DNA у вас доступна для изменения из вне.

Cell *c = [[Cell init] alloc];
c.DNA = [NSMutableArray array]; // разрешенная операция
[c.DNA removeAllObjects]; // разрешенная операция

Определив сеттер извне она больше не меняется.

hmmvot replied:
Извините, но я удалил это сообщение сразу после отправки.  Не правильно понял ваш код сначала, этот комментарий не актуален.

hmmvot commented on DNA/Cell.m:
```diff
@@ -0,0 +1,39 @@
+    return self;
+}
+-(NSArray*) possibleNucleotides {
+    return [NSArray arrayWithObjects:@"A",@"T",@"G",@"C", nil];
```
Вы ничего не сэкономили - каждый раз возвращаете новый объект. Используйте спецификатор static.

(Если захотите хранить NSArray в статической переменной - обратите внимание на метод класса - (void) initialize. Но здесь лучше использовать обычный массив C)

l0rda replied:
вынес в дефайн в последнем коммите

l0rda committed:
улучшение

массив с нуклеотидами вынес в дефайн
убрал определение description из интерфейса, убрал сеттер setDNA

l0rda committed:
исправлена логическая ошибка с distance

hmmvot commented on DNA/Cell.h:
```diff
@@ -0,0 +1,20 @@
+#import <Foundation/Foundation.h>
+#define DNA_LEN     100
+#define NUKE_LEN    4
+#define NUKE_ARR    @[@"A",@"T",@"G",@"C"]
```
Я, наверное, вас уже замучил, но такой вариант ничего не изменил.) Теперь при каждом упоминании в коде NUKE_ARR у вас будет создаваться новый массив.

Вы можете использовать такое решение:

``` objective-c
// Cell.h
@interface Cell
+ (NSArray *) letters;
@end

// Cell.m
@implementation Cell

static NSArray *lettersArray;

+ (void) initialise
{
    lettersArray = [[NSArray alloc] initWithObjects: @"A", @"T", @"G", @"C", nil];
}

+ (NSArray *) letters
{
    return lettersArray;
}

@end

// AnyOtherSource.m
NSLog(@"%@", [Cell letters]);
```

Или посмотреть как это сделал я с помощью спецификатора extern и C-массива.

hmmvot replied:
Про метод initialize класса NSObject:
https://developer.apple.com/library/mac/#documentation/Cocoa/Reference/Foundation/Classes/NSObject_Class/Reference/Reference.html#//apple_ref/occ/clm/NSObject/initialize

Про static (и extern):
http://www.rsdn.ru/forum/cpp/2706624.1

Общий смысл таков:
1) + (void) initialize - это метод класса, а не объекта класса, и он будет вызван только однажды (до того, как вы пошлете классу Cell первое сообщение).
2) Статическая переменная создается в только однажды.

l0rda replied:
Да,я уже понял что за глупость написал и удалил коммент :)
Сделал коммит, надеюсь теперь все в порядке.

l0rda committed:
Исправление

массив с нуклеотидами теперь в классе, появляется через initialize()

hmmvot commented on DNA/Cell.h:
```diff
@@ -0,0 +1,20 @@
+#define DNA_LEN     100
+#define NUKE_LEN    4
+
+extern NSMutableArray *nucleotides;
```
Вот от этого не забудьте избавиться.

l0rda replied:
спасибо, проглядел

l0rda committed:
оптимизация

убрал старый неиспользуемый код

l0rda committed:
еще один вариант мутации

основан на кратности текущего индекса нуклеотида

l0rda committed:
улучшение

точное вычисление кол-ва жлементов массива DNA для замены, если в
массиве больше 100 элементов

l0rda committed:
FINAL

хотелось еще поэксперементировать, но времени нет что-то еще править,
по ТЗ все работает :)

hmmvot commented on DNA/Cell+mutator.m:
```diff
@@ -0,0 +1,47 @@
+
+@implementation Cell(mutator)
+-(void)mutate: (int) mutatePercent {
+    NSAssert(mutatePercent < 100 && mutatePercent > 0, @"invalid percent");
```
Программа падает при mutatePercent == 0 и mutatePercent == 100.

l0rda replied:
при 0 должна падать, при 100 не должна, этот поправил у себя

l0rda replied:
хотя, сделал чтобы при 0 просто выходила

hmmvot commented on DNA/Cell+mutator.m:
```diff
@@ -0,0 +1,47 @@
+-(void)mutate: (int) mutatePercent {
+    NSAssert(mutatePercent < 100 && mutatePercent > 0, @"invalid percent");
+    // вычисляем примерное кол-во индексов для замены, если кол-во элементов DNA != 100
+    int numberIndexesToReplace = (float)DNA_LEN / 100 * mutatePercent;
```
Зачем нам выполнять все начиная с этой строки, если mutatePercent == 0?

l0rda replied:
согласен

l0rda replied:
а оно и не выполняется, а выпадает assert

hmmvot commented on DNA/Cell+mutator.m:
```diff
@@ -0,0 +1,47 @@
+    // вычисляем примерное кол-во индексов для замены, если кол-во элементов DNA != 100
+    int numberIndexesToReplace = (float)DNA_LEN / 100 * mutatePercent;
+    
+    static NSMutableArray *mutateIndexArray, *dnaIndexArray;
```
Наличие спецификатора static здесь не оправданно (вы не пользуетесь тем, что mutateIndexArray и dnaIndexArray сохраняют свои значения между вызовами метода mutate).

l0rda replied:
согласен

hmmvot commented on DNA/Cell+mutator.m:
```diff
@@ -0,0 +1,47 @@
+    int numberIndexesToReplace = (float)DNA_LEN / 100 * mutatePercent;
+    
+    static NSMutableArray *mutateIndexArray, *dnaIndexArray;
+    mutateIndexArray = [[NSMutableArray alloc] initWithCapacity:mutatePercent];
```
Почему предполагаемый размер массива равен mutatePercen, а не numberIndexesToReplace?

l0rda replied:
Это ошибка :(

hmmvot commented on DNA/Cell+mutator.m:
```diff
@@ -0,0 +1,47 @@
+        numberIndexesToReplace--;
+    }
+    
+    for (id mIndex in mutateIndexArray) {
```
Было бы хорошо более строго указать тип mIndex (вы же знаете, что это NSNumber).

l0rda replied:
Пока не до конца разобрался с типами в obj-c. Исправлюсь по мере развития :)

hmmvot commented on DNA/Cell+mutator.m:
```diff
@@ -0,0 +1,47 @@
+        // чтобы нуклеотид в позиции был гарантированно заменен на другой
+        NSString *currentNuke = [DNA objectAtIndex:[mIndex integerValue]];
+        NSUInteger current_nuke_index = [[Cell nucleotides] indexOfObject:currentNuke];
+        id newNucleotide;
```
Здесь тоже лучше бы поменять id на NSString.

l0rda replied:
Пока не до конца разобрался с типами в obj-c. Исправлюсь по мере развития :)

hmmvot commented on DNA/Cell+mutator.m:
```diff
@@ -0,0 +1,47 @@
+            newNucleotide = [[Cell nucleotides] objectAtIndex:++current_nuke_index];
+        } else {
+            newNucleotide = [[Cell nucleotides] objectAtIndex:--current_nuke_index];
+        }
```
Из элемента A можно получить только элемент T (например). Мутация явно не рандомная.)

l0rda replied:
Это я согласен, но есть чем возразить )во-первых, тут были две другие реализации метода, полностью рандомные, но вы сказали, что это интереснее ) во-вторых мутация никогда не бывает рандомная, все подчиняется своим законам ;). Согласен в общем.

l0rda replied:
Посмотрите пожалуйста новый вариант :)

hmmvot commented:
8.5
-1.5 балла за метод mutate (не работает для входных значений 0 и 100, мутация не рандомная).
Недочеты указал в комментариях к коду.

l0rda committed:
fix

Исправление ошибки с неверным размером массива с индексами для замены.
Более корректное объявление переменных.

l0rda committed:
новый вариант мутации нуклеотидов

Добавлен статический массив массивов с индексами нуклеотидов. Так как
нуклеотида 4, то 4 массива, в 0-ом нет индекса 0, в первом нет индекса
1 и т.д. При мутации берем текущий индекс нуклеотида, по нему выбираем
массив с 3 возможными индексами и делаем по ним случайную выборку(1 из
3 элементов). Таким образом, при проходе каждого элемента массива DNA
мы не создаем никаких новых массивов.

l0rda committed:
удалили лишние комментарии

l0rda committed:
еще более красивая мутация

получаем новый нуклеотид в классе

hmmvot commented on DNA/Cell.m:
```diff
@@ -0,0 +1,63 @@
+    return noIndex;
+}
+
++ (NSString *) getRandomNucleotideWithoutIndex: (NSUInteger) curIndex
```
Хитро.)
Старайтесь не использовать "магических чисел". 3 == (NUKE_LEN - 1).

l0rda replied:
Ваша правда. Но в данном конкретном случае я долго думал :) и надумал раз уж я задаю сам эти массивы и точно знаю их длину и что она не изменится, то почему бы и нет.

l0rda replied:
Поправил.

l0rda committed:
фиксим машшшические числа и убираем лишний код
